File: my-solutions/0230-kth-smallest-element-in-a-bst/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        int count=0;
        TreeNode* cur=root;
        int result=-1;
        while(cur!=NULL){
            if(cur->left==NULL){
                count++;
                if(count==k){
                    result = cur->val;
                }
                cur = cur->right;   
            }else{
                TreeNode* prev=cur->left;
                while(prev->right && prev->right!=cur){
                    prev=prev->right;
                }
                if(prev->right==NULL){
                    prev->right=cur;
                    cur=cur->left;
                }else{
                    prev->right=NULL;
                    count++;
                    if(count==k){
                        result = cur->val;
                    }
                    cur=cur->right;
                }
            }
        }
        return result;
    }
};

File: my-solutions/0230-kth-smallest-element-in-a-bst/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        traversal = []
        self.helper(root, k, traversal)
        return traversal[-1]
    

    def helper(self, node: Optional[TreeNode], k: int, traversal: List[int]) -> None:
        if len(traversal) == k:
            return
        elif not node:
            return
        
        self.helper(node.left, k, traversal)
        if len(traversal) == k:
            return
        
        traversal.append(node.val)
        if len(traversal) == k:
            return
        
        self.helper(node.right, k, traversal)
